From 1a771bb318a84f02803728bee695ebf735f4e3a7 Mon Sep 17 00:00:00 2001
From: Prithvi A Pai <prithvi.a.pai@intel.com>
Date: Tue, 10 Oct 2023 12:24:55 +0530
Subject: [PATCH] FIPS Add route in Managers

This patch helps FIPS enabling/disabling from REST interface.

GET: /redfish/v1/Managers/bmc/SecurityPolicy
GET on the SecurityPolicy exposes the Intel OEM properties which provides the following information.

AvailableProviders - Contains all the available FIPS provider
Enabled - Tells if FIPS is Enabled/Disabled
Version - Tells the version of FIPS provider
BMCResetRequired - Tells if BMC reboot is needed for installation of
FIPS.
PATCH: /redfish/v1/Managers/bmc/SecurityPolicy
PATCH on SecurityPolicy helps to Enable/Disable FIPS

Tested:
Tried Enabling/Disabling FIPS and it works fine

Signed-off-by: Prithvi A Pai <prithvi.a.pai@intel.com>
---
 redfish-core/include/registries/privilege_registry.hpp | 4 ++++
 redfish-core/lib/managers.hpp                          | 2 ++
 2 files changed, 6 insertions(+)

diff --git a/redfish-core/include/registries/privilege_registry.hpp b/redfish-core/include/registries/privilege_registry.hpp
index 57e54649..653ed0a5 100644
--- a/redfish-core/include/registries/privilege_registry.hpp
+++ b/redfish-core/include/registries/privilege_registry.hpp
@@ -1657,5 +1657,9 @@ const static auto& putLicense = privilegeSetConfigureComponents;
 const static auto& deleteLicense = privilegeSetConfigureComponents;
 const static auto& postLicense = privilegeSetConfigureComponents;
 
+// FIPS Enablement
+const static auto& getFIPS = privilegeSetLogin;
+const static auto& patchFIPS = privilegeSetConfigureComponents;
+
 } // namespace redfish::privileges
 // clang-format on
diff --git a/redfish-core/lib/managers.hpp b/redfish-core/lib/managers.hpp
index 218124d0..1e8120a5 100644
--- a/redfish-core/lib/managers.hpp
+++ b/redfish-core/lib/managers.hpp
@@ -1980,6 +1980,8 @@ inline void requestRoutesManager(App& app)
             "/redfish/v1/Managers/bmc/NetworkProtocol";
         asyncResp->res.jsonValue["EthernetInterfaces"]["@odata.id"] =
             "/redfish/v1/Managers/bmc/EthernetInterfaces";
+        asyncResp->res.jsonValue["SecurityPolicy"]["@odata.id"] =
+            "/redfish/v1/Managers/bmc/SecurityPolicy";
 
 #ifdef BMCWEB_ENABLE_VM_NBDPROXY
         asyncResp->res.jsonValue["VirtualMedia"]["@odata.id"] =
-- 
2.34.1

